lib-src/libcs/openp.3
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1994 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" See <cmu_copyright.h> for use and distribution information.
.\" 
.\" 
.\" HISTORY
.\" $Log: openp.3,v $
.\" Revision 1.1  1996/11/22 19:19:36  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:55  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:48  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:55:22  moore]
.\"
.\" $EndLog$
.\" Copyright (c) 1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
.\" SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Users of this software agree to return to Carnegie Mellon any
.\" improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" Export of this software is permitted only after complying with the
.\" regulations of the U.S. Deptartment of Commerce relating to the
.\" Export of Technical Data.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" HISTORY
.\" $Log: openp.3,v $
.\" Revision 1.1  1996/11/22 19:19:36  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:55  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:48  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:55:22  moore]
.\"
.\" Revision 1.1.1.2  1995/10/18  14:55:22  moore
.\" 	Created libcs man(3) pages from mach entries
.\"
.\" Revision 1.2  90/12/12  15:55:15  mja
.\" 	Add copyright/disclaimer for distribution.
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\" 	Revised for 4.3.
.\" 
.\" 24-Mar-86  Rudy Nedved (ern) at Carnegie-Mellon University
.\" 	Updated for 4.2BSD.
.\" 
.\" 05-Dec-79  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Created.
.\" 
.TH OPENP 3 3/24/86
.CM 4
.SH "NAME"
openp, fopenp \- search for file and open it
.SH "SYNOPSIS"
.B 
#include <sys/file.h>
.br
.B 
int openp (searchlist, path, buffer, flags, mode);
.br
.B 
char *searchlist, *path, *buffer;
.br
.B 
int flags, mode;

.B 
#include <sys/file.h>
.br
.B 
#include <stdio.h>
.br
.B 
FILE *fopenp (searchlist,path,buffer,type);
.br
.B 
char *searchlist, *path, *buffer ,*type;
.SH "DESCRIPTION"
.I 
Openp
and
.I 
fopenp
use
.IR searchp (3)
to search for a file, and open the file when and if it is
found.
The value returned will be the normal value of
.IR open (2)
or
.IR fopen (3s)
-- a file descriptor or FILE pointer on success, and a
\-1 or a 0 on failure.

.I 
Searchlist
is assumed to be a list of directory path names separated by colons;
one by one, these names are parsed and concatenated (with a
separating slash) onto
.I 
path
to form a complete file path name.
An attempt is then made to
open the file with this name; if successful,
.I 
openp
and
.I 
fopenp
return with a success indication; otherwise, searching
continues.
If success is achieved, then the resulting
pathname is copied into the string
.I 
buffer,
provided by the user.
The
.I 
open
and
.I 
fopen
calls are made with the appropriate
.I 
mode, flags
and
.I 
type
as specified by the parameter. The 
.IR open(2)
and
.IR fopen(3S)
define what
.I 
mode, flags
and
.I 
type
should be respectively.
.SH "LIBRARY NOTES"
You must link against /usr/local/lib/libcs.a to define this routine.
.SH "SEE ALSO"
.IR searchp (3), 
.IR open (2), 
.IR fopen (3s), 
.IR wantread (3), 
.IR wantwrite (3),
.IR fwantread (3), 
.IR fwantwrite (3)
.SH "DIAGNOSTICS"
.I 
Openp
returns \-1 on error (no openable file found);
.I 
fopenp
returns 0.


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode


lib-src/libcs/openp.c
#ifndef _BLURB_
#define _BLURB_
/*
 * This code was originally part of the CMU SCS library "libcs".
 * A branch of that code was taken in May 1996, to produce this
 * standalone version of the library for use in Coda and Odyssey
 * distribution.  The copyright and terms of distribution are
 * unchanged, as reproduced below.
 *
 * Copyright (c) 1990-96 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
 * SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Users of this software agree to return to Carnegie Mellon any
 * improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 *
 * Export of this software is permitted only after complying with the
 * regulations of the U.S. Deptartment of Commerce relating to the
 * Export of Technical Data.
 */

static char *rcsid = "$Header: openp.c,v 1.1 96/06/03 19:00:56 satya Exp $";
#endif /*_BLURB_*/

/*  openp, fopenp  --  search pathlist and open file
 *
 *  Usage:
 *	i = openp (path,file,complete,flags,mode)
 *	f = fopenp (path,file,complete,type)
 *	int i,flags,mode;
 *	FILE *f;
 *	char *path,*file,*complete,*type;
 *
 *  Openp searches for "file" in the pathlist "path";
 *  when the file is found and can be opened by open()
 *  with the specified "flags" and "mode", then the full filename
 *  is copied into "complete" and openp returns the file
 *  descriptor.  If no such file is found, openp returns -1.
 *  Fopenp performs the same function, using fopen() instead
 *  of open() and type instead of flags/mode; it returns 0 if no
 *  file is found.
 *
 */

#include <stdio.h>

#include "libcs.h"

static int flgs,mod,value;
static const char *ftyp;
static FILE *fvalue;

static int func (fnam)
char *fnam;
{
	value = open (fnam,flgs,mod);
	return (value < 0);
}

static int ffunc (fnam)
char *fnam;
{
	fvalue = fopen (fnam,ftyp);
	return (fvalue == 0);
}

int openp (path,file,complete,flags,mode)
const char *path, *file;
char *complete;
int flags, mode;
{
	flgs = flags;
	mod = mode;
	if (searchp(path,file,complete,func) < 0)  return (-1);
	return (value);
}

FILE *fopenp (path,file,complete,ftype)
const char *path, *file, *ftype;
char *complete;
{
	ftyp = ftype;
	if (searchp(path,file,complete,ffunc) < 0)  return (0);
	return (fvalue);
}
